pvlib/iotools/solcast.py

```diff
@@ -0,0 +1,443 @@
+""" Functions to access data from the Solcast API.
+"""
+
+import requests
+import pandas as pd
+from dataclasses import dataclass
+
+
+BASE_URL = "https://api.solcast.com.au/data"
+
+
+@dataclass
+class ParameterMap:
+    solcast_name: str
+    pvlib_name: str
+    conversion: callable = lambda x: x
+
+
+# define the conventions between Solcast and PVLib nomenclature and units
+VARIABLE_MAP = [
+    # air_temp -> temp_air (deg C)
+    ParameterMap("air_temp", "temp_air"),
+    # surface_pressure (hPa) -> pressure (Pa)
+    ParameterMap("surface_pressure", "pressure", lambda x: x*100),
+    # dewpoint_temp -> temp_dew (deg C)
+    ParameterMap("dewpoint_temp", "temp_dew"),
+    # gti (W/m^2) -> poa_global (W/m^2)
+    ParameterMap("gti", "poa_global"),
+    # wind_speed_10m (m/s) -> wind_speed (m/s)
+    ParameterMap("wind_speed_10m", "wind_speed"),
+    # wind_direction_10m (deg) -> wind_direction  (deg)
+    ParameterMap("wind_direction_10m", "wind_direction"),
+    # azimuth -> solar_azimuth (degrees) (different convention)
+    ParameterMap(
+        "azimuth", "solar_azimuth", lambda x: abs(x) if x <= 0 else 360 - x
+    ),
+    # precipitable_water (kg/m2) -> precipitable_water (cm)
+    ParameterMap("precipitable_water", "precipitable_water", lambda x: x*10),
+    # zenith -> solar_zenith
+    ParameterMap("zenith", "solar_zenith")
+]
+
+
+def get_solcast_tmy(
+    latitude, longitude, api_key, map_variables=True, **kwargs
+):
+    """Get the irradiance and weather for a
+    Typical Meteorological Year (TMY) at a requested location.
+
+    Derived from satellite (clouds and irradiance over
+    non-polar continental areas) and numerical weather models (other data).
+    The TMY is calculated with data from 2007 to 2023.
+
+    Parameters
+    ----------
+    latitude : float
+        in decimal degrees, between -90 and 90, north is positive
+    longitude : float
+        in decimal degrees, between -180 and 180, east is positive
+    api_key : str
+        To access Solcast data you will need an API key [1]_.
+    map_variables: bool, default: True
+        When true, renames columns of the DataFrame to pvlib variable names
+        where applicable. See variable :const:`VARIABLE_MAP`.
+    kwargs:
+        Optional parameters passed to the API.
+        See [2]_ for full list of parameters.
+
+    Returns
+    -------
+    data : pandas.DataFrame
+        containing the values for the parameters requested.The times
+        in the DataFrame index indicate the midpoint of each interval.
+
+    Examples
+    --------
+    >>> pvlib.iotools.solcast.get_solcast_tmy(

```

Comment:
```suggestion
    >>> df, meta = pvlib.iotools.solcast.get_solcast_tmy(
```

And same for the other examples